File: benchmarks/reference/go/fannkuch_redux.go
package main

import (
	"fmt"
	"time"
)

func fannkuch(n int) int {
	perm := make([]int, n)
	perm1 := make([]int, n)
	count := make([]int, n)

	for i := 0; i < n; i++ {
		perm1[i] = i
	}
	for i := 0; i < n; i++ {
		count[i] = i + 1
	}

	maxFlips := 0
	checksum := 0
	permCount := 0

	for {
		if perm1[0] != 0 {
			copy(perm, perm1)
			flips := 0
			for perm[0] != 0 {
				k := perm[0]
				for lo, hi := 0, k; lo < hi; lo, hi = lo+1, hi-1 {
					perm[lo], perm[hi] = perm[hi], perm[lo]
				}
				flips++
			}
			if flips > maxFlips {
				maxFlips = flips
			}
			if permCount%2 == 0 {
				checksum += flips
			} else {
				checksum -= flips
			}
		}
		permCount++

		r := 1
		for ; r < n; r++ {
			perm0 := perm1[0]
			for i := 0; i < r; i++ {
				perm1[i] = perm1[i+1]
			}
			perm1[r] = perm0
			count[r]--
			if count[r] > 0 {
				break
			}
			count[r] = r + 1
		}
		if r >= n {
			break
		}
	}

	fmt.Printf("checksum: %d\n", checksum)
	return maxFlips
}

func main() {
	n := 10
	start := time.Now()
	result := fannkuch(n)
	ms := time.Since(start).Milliseconds()
	fmt.Printf("max flips: %d\n", result)
	fmt.Printf("elapsed: %d ms\n", ms)
}

File: benchmarks/reference/python/fannkuch_redux.py
import time

def fannkuch(n):
    perm1 = list(range(n))
    count = [i + 1 for i in range(n)]
    max_flips = 0
    checksum = 0
    perm_count = 0

    while True:
        if perm1[0] != 0:
            perm = perm1[:]
            flips = 0
            while perm[0] != 0:
                k = perm[0]
                perm[:k+1] = perm[:k+1][::-1]
                flips += 1
            if flips > max_flips:
                max_flips = flips
            if perm_count % 2 == 0:
                checksum += flips
            else:
                checksum -= flips
        perm_count += 1

        r = 1
        while r < n:
            perm0 = perm1[0]
            for i in range(r):
                perm1[i] = perm1[i + 1]
            perm1[r] = perm0
            count[r] -= 1
            if count[r] > 0:
                break
            count[r] = r + 1
            r += 1
        if r >= n:
            break

    print(f"checksum: {checksum}")
    return max_flips

if __name__ == "__main__":
    n = 10
    t0 = time.monotonic_ns()
    result = fannkuch(n)
    ms = (time.monotonic_ns() - t0) // 1_000_000
    print(f"max flips: {result}")
    print(f"elapsed: {ms} ms")
